gr-blocks/lib/selector_impl.cc

```diff
@@ -0,0 +1,141 @@
+#include <string.h>
+#include <stdexcept>
+
+namespace gr {
+namespace blocks {
+
+selector::sptr
+selector::make(size_t itemsize, unsigned int input_index, unsigned int output_index)
+{
+    return gnuradio::get_initial_sptr(
+        new selector_impl(itemsize, input_index, output_index));
+}
+
+selector_impl::selector_impl(size_t itemsize,
+                             unsigned int input_index,
+                             unsigned int output_index)
+    : block("selector",
+            io_signature::make(1, -1, itemsize),
+            io_signature::make(1, -1, itemsize)),
+      d_itemsize(itemsize),
+      d_enabled(true),
+      d_input_index(input_index),
+      d_output_index(output_index),
+      d_num_inputs(0),
+      d_num_outputs(0)
+{
+    message_port_register_in(pmt::mp("en"));
+    set_msg_handler(pmt::mp("en"), [this](pmt::pmt_t msg) { this->handle_enable(msg); });
+
+    // TODO: add message ports for input_index and output_index
+}
+
+selector_impl::~selector_impl() {}
+
+void selector_impl::set_input_index(unsigned int input_index)
+{
+    gr::thread::scoped_lock l(d_mutex);
+    if (input_index < d_num_inputs)
+        d_input_index = input_index;
+    else
+        throw std::out_of_range("input_index must be < ninputs");
+}
+
+void selector_impl::set_output_index(unsigned int output_index)
+{
+    gr::thread::scoped_lock l(d_mutex);
+    if (output_index < d_num_outputs)
+        d_output_index = output_index;
+    else
+        throw std::out_of_range("output_index must be < noutputs");
+}
+
+void selector_impl::handle_enable(pmt::pmt_t msg)
+{
+    if (pmt::is_bool(msg)) {
+        bool en = pmt::to_bool(msg);
+        gr::thread::scoped_lock l(d_mutex);
+        d_enabled = en;
+    } else {
+        GR_LOG_WARN(d_logger,
+                    "handle_enable: Non-PMT type received, expecting Boolean PMT");
+    }
+}
+
+void selector_impl::forecast(int noutput_items, gr_vector_int& ninput_items_required)
+{
+    unsigned ninputs = ninput_items_required.size();
+    for (unsigned i = 0; i < ninputs; i++) {
+        ninput_items_required[i] = noutput_items;
+    }
+}
+
+bool selector_impl::check_topology(int ninputs, int noutputs)
+{
+    if ((int) d_input_index < ninputs && (int) d_output_index < noutputs) {
+        d_num_inputs = (unsigned int)ninputs;
+        d_num_outputs = (unsigned int)noutputs;
+        return true;
+    } else {
+        GR_LOG_WARN(d_logger,
+                    "check_topology: Input or Output index greater than number of ports");
+        return false;
+    }
+}
+
+int selector_impl::general_work(int noutput_items,
+                                gr_vector_int& ninput_items,
+                                gr_vector_const_void_star& input_items,
+                                gr_vector_void_star& output_items)
+{
+    const uint8_t** in = (const uint8_t**)&input_items[0];
+    uint8_t** out = (uint8_t**)&output_items[0];
+
+    gr::thread::scoped_lock l(d_mutex);
```

Comment:
See my previous comment about the mutex. `work()` is in the hot loop, and this gives us some safety against... what exactly? whereas it adds locking load.

Reply:
Seems like @marcusmueller was fairly emphatic about the mutex - if it's safer to leave it out I'll delete those lines

Reply:
@marcusmueller Has the last word here. It's OK. It's safer to *not* leave it out, my argument is just that we don't get anything by adding that safety, and I don't think anyone can show my an app that works with it and not without it.